index/indexer.go

```diff
@@ -644,12 +651,77 @@ func (e *indexer) emitItemOfReferences(outV string, inVs []string, docID string)
 	return id, e.emit(protocol.NewItemOfReferences(id, outV, inVs, docID))
 }
 
-func (e *indexer) emitMoniker(kind, sourceID, identifier string) error {
-	monikerID := e.nextID()
-	err := e.emit(protocol.NewMoniker(monikerID, kind, protocol.LanguageID, identifier))
+func (e *indexer) emitImportMoniker(sourceID, identifier string) error {
+	for _, moduleName := range packagePrefixes(strings.Split(identifier, ":")[0]) {
+		moduleVersion, ok := e.dependencies[moduleName]
+		if !ok {
+			continue
+		}
+
+		packageInformationID, err := e.ensurePackageInformation(moduleName, moduleVersion)
+		if err != nil {
+			return err
+		}
+
+		return e.addMonikers("import", e.nextID(), sourceID, packageInformationID)
+	}
+
+	return nil
+}
+
+func (e *indexer) emitExportMoniker(sourceID, identifier string) error {
+	packageInformationID, err := e.ensurePackageInformation(e.moduleName, e.moduleVersion)
+	if err != nil {
+		return err
+	}
+
+	return e.addMonikers("export", e.nextID(), sourceID, packageInformationID)
+}
+
+func (e *indexer) ensurePackageInformation(packageName, version string) (string, error) {
+	packageInformationID, ok := e.packageInformationIDs[packageName]
+	if !ok {
+		packageInformationID = uuid.New().String()
+		err := e.emit(protocol.NewPackageInformation(packageInformationID, packageName, "gomod", version))
+		if err != nil {
+			return "", err
+		}
+
+		e.packageInformationIDs[packageName] = packageInformationID
+	}
+
+	return packageInformationID, nil
+}
+
+// addMonikers outputs a "gomod" moniker vertex, attaches the given package vertex
+// identifier to it, and attaches the new moniker to the source moniker vertex.
+func (e *indexer) addMonikers(kind string, identifier string, sourceID, packageID string) error {
+	monikerID := uuid.New().String()
+	err := e.emit(protocol.NewMoniker(monikerID, kind, "gomod", identifier))
+	if err != nil {
+		return err
+	}
+
+	err = e.emit(protocol.NewPackageInformationEdge(uuid.New().String(), monikerID, packageID))
 	if err != nil {
 		return err
 	}
 
-	return e.emit(protocol.NewMonikerEdge(e.nextID(), sourceID, monikerID))
+	err = e.emit(protocol.NewNextMonikerEdge(uuid.New().String(), sourceID, monikerID))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func packagePrefixes(packageName string) []string {
```

Comment:
Would you mind add a docstring for this function and include some example input/outputs?